giraph-core/src/main/java/org/apache/giraph/ooc/policy/MemoryEstimatorOracle.java

```diff
@@ -0,0 +1,851 @@
+          oocBytesInjected == last[4]) {
+          if (LOG.isDebugEnabled()) {
+            LOG.debug(
+              "addRecord: avoiding to add the same entry as the last one!");
+          }
+          return;
+        }
+      }
+      dataSamples.add(new double[] {edges, vertices, verticesProcessed,
+        bytesReceived, oocBytesInjected});
+      memorySamples.add((double) memUsed);
+
+      // Weed out the columns that are all zero
+      validColumnIndices.clear();
+      for (int i = 0; i < 5; ++i) {
+        boolean validIndex = false;
+        // Check if there is a non-zero entry in the column
+        for (double[] value : dataSamples) {
+          if (value[i] != 0) {
+            validIndex = true;
+            break;
+          }
+        }
+        if (validIndex) {
+          // check if all entries are not equal to each other
+          double firstValue = -1;
+          boolean allEqual = true;
+          for (double[] value : dataSamples) {
+            if (firstValue == -1) {
+              firstValue = value[i];
+            } else {
+              if (Math.abs((value[i] - firstValue) / firstValue) > 0.01) {
+                allEqual = false;
+                break;
+              }
+            }
+          }
+          validIndex = !allEqual;
+          if (validIndex) {
+            // Check if the column has linear dependency with another column
+            for (int col = i + 1; col < 5; ++col) {
+              if (isLinearDependence(dataSamples, i, col)) {
+                validIndex = false;
+                break;
+              }
+            }
+          }
+        }
+
+        if (validIndex) {
+          validColumnIndices.add(i);
+        }
+      }
+
+      // If we filtered out columns in the previous step, we are going to run
+      // the regression without those columns.
+
+      // Create the coefficient table
+      boolean setIsValid = false;
+      lock.lock();
+      try {
+        if (validColumnIndices.size() >= 1 &&
+          dataSamples.size() >= validColumnIndices.size() + 1) {
+
+          double[][] xValues = new double[dataSamples.size()][];
+          double[] yValues = new double[memorySamples.size()];
+          fillXMatrix(dataSamples, validColumnIndices, xValues);
+          copyVectorToArray(memorySamples, yValues);
+          mlr.newSampleData(yValues, xValues);
+          calculateRegression(coefficient, validColumnIndices, mlr);
+
+          // After the computation of the regression, some coefficients may have
+          // values outside the valid value range. In this case, we set the
+          // coefficient to the minimum or maximum value allowed, and re-run the
+          // regression.
+          boolean changed;
+          extreme[3] = -1;
+          extreme[4] = -1;
+          do {
+            changed = refineCoefficient(4, 1, 2, xValues, yValues);
+            changed |= refineCoefficient(3, 0, 2, xValues, yValues);
```

Comment:
What are these valid ranges? Please, add an explanation. 
oocBytes coefficient has to be between 1 and 2, why?
receivedBytes has to be between 0 and 2, again why?